### ISSUE TITLE
Add mlir-vulkan-runner

### ISSUE THREAD

**Issue by antiagainst** [tensorflow/mlir#60]
It would be nice to have a simple testing harness to exercise the SPIR-V blob generated and serialized from SPIR-V CodeGen on Vulkan-capable devices, similar to mlir-cuda-runner.

Ideally the runner should inspect the SPIR-V module to help building the resource interface for a Vulkan compute pipeline, run the compute pipeline by feeding in test data and then check the result data.

**Comment by denis0x0D** [comment]
@antiagainst if noone is working on it, I can take it, I've already have a kinda "Vulkan runtime" to test spir-v compute shaders generated by hands written SPIR-V IRBuilder, I learned it from the blog of Neil Henning http://www.duskborn.com/posts/a-simple-vulkan-compute-example/ and code is mostly the same.

**Comment by antiagainst** [comment]
Please feel free to proceed if you'd like to take on it! Thanks a lot, @denis0x0D!

**Comment by denis0x0D** [comment]
@antiagainst thanks!

**Comment by MaheshRavishankar** [comment]
Fantastic! Thanks Denis!

**Comment by denis0x0D** [comment]
@antiagainst @MaheshRavishankar can you please comment on the following, to make sure I understand the specifications of the task in right way,  and please fix me if I'm wrong.

From the high-level view the mlir-vulkan-runner must consits of 2 main parts:
compile-time passes and runtime machinery.

1. Pipeline of the passes:
a. The pass which outlines gpu.launch body to kernel function. (GpuKernelOutLiningPass)
b. The pass which converts GPU dialect into SPIR-V dialect. (GPUToSPIRVPass)
c. The codegen stage which generates spir-v binary from spir-v kernel. (SPIR-V serialization) is similiar to ConvertGPUKernelToCubinPass.
d. The pass which replaces "gpu.launch_function" with call into runtime part, and instruments code with calls to runtime machinery simliar to GpuLanchFunctToCudaCallsPass does.

2. Runtime machinery:
a. Consists of the interceptors which must use Vulkan API.
b. Takes responsibilities for the memory management, buffer creation, device creation, kernel launch and so on.

So, in this case I can reuse existing mlir::JitRunnerMain which uses mlir::ExecuteEngine to run the program.
For example the representation of the module after GPU->SPIRV converter pass.

```
module {
  spv.module "Logical" "VulkanKHR" {
    %0 = spv.Variable bind(0, 0) : !spv.ptr<f32, StorageBuffer>
    %1 = spv.Variable bind(0, 1) : !spv.ptr<!spv.array<12 x f32>, StorageBuffer>
    func @kernel_1(%arg0: !spv.ptr<f32, StorageBuffer>, %arg1: !spv.ptr<!spv.array<12 x f32>, StorageBuffer>) {
      spv.Return
    }
    spv.EntryPoint "GLCompute" @kernel_1, %0, %1 : !spv.ptr<f32, StorageBuffer>, !spv.ptr<!spv.array<12 x f32>, StorageBuffer>
  }
  func @kernel_1(%arg0: f32, %arg1: memref<12xf32, 1>)
  attributes  {gpu.kernel} {
    return
  }
  func @main() {
    %0 = "op"() : () -> f32
    %1 = "op"() : () -> memref<12xf32, 1>
    %c1 = constant 1 : index
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %0, %1) {kernel = @kernel_1} : (index, index, index, index, index, index, f32, memref<12xf32, 1>) -> ()
    return
  }
}

```
On the next stage I should generate spirv binary from spv.module and kernel,

```
; SPIR-V
; Version: 1.0
; Generator: Khronos; 22
; Bound: 14
; Schema: 0
               OpMemoryModel Logical VulkanKHR
               OpEntryPoint GLCompute %kernel_1 "kernel_1" %3 %8
               OpName %kernel_1 "kernel_1"
               OpDecorate %3 Binding 0
               OpDecorate %3 DescriptorSet 0
               OpDecorate %8 Binding 1
               OpDecorate %8 DescriptorSet 0
      %float = OpTypeFloat 32
%_ptr_StorageBuffer_float = OpTypePointer StorageBuffer %float
        %int = OpTypeInt 32 1
     %int_12 = OpConstant %int 12
%_arr_float_int_12 = OpTypeArray %float %int_12
%_ptr_StorageBuffer__arr_float_int_12 = OpTypePointer StorageBuffer %_arr_float_int_12
       %void = OpTypeVoid
          %9 = OpTypeFunction %void %_ptr_StorageBuffer_float %_ptr_StorageBuffer__arr_float_int_12
          %3 = OpVariable %_ptr_StorageBuffer_float StorageBuffer
          %8 = OpVariable %_ptr_StorageBuffer__arr_float_int_12 StorageBuffer
   %kernel_1 = OpFunction %void None %9
         %12 = OpFunctionParameter %_ptr_StorageBuffer_float
         %13 = OpFunctionParameter %_ptr_StorageBuffer__arr_float_int_12
               OpReturn
               OpFunctionEnd

```
On next stage I should instrument function "main" with calls into runtime part. 
For example instrumented function main with external call: 

```
...
call mcuMemHostRegister(%memref_type)
...

```
Runtime call:
```
void mcuMemHostRegister(const memref_t arg) {
...
VkDeviceMemory memory;
vkAllocateMemory(device, &memoryAllocateInfo, 0, &memory);
vkMapMemory(device, memory, 0, arg.length * sizeof(float), 0, (void **)&host_memory_ptr);
VkBuffer buffer;
vkCreateBuffer(device, &bufferCreateInfo, 0, &buffer);
vkBindBufferMemory(device, buffer, memory, 0);
...

```
and so on.

This is high level view and I can miss a lot, please fix me if I'm wrong.
Thanks.

**Comment by MaheshRavishankar** [comment]
Thanks Denis! I can comment with more confidence on the compiler parts. I am fairly sure there are still missing parts in the GPU -> SPIRV conversion. I am going to look into them with some realistic examples. If you find any issues, please do point out on this bug (so that we don't end up duplicating the work).
Overall though the approach you listed makes sense to me. You should probably wait for @antiagainst to weigh in as well.

**Comment by antiagainst** [comment]
Hey @denis0x0D,

One thing to keep in mind regarding this runner is that it is for testing purpose; it is not intended to be a full-fledged performant Vulkan runtime that does all the smarts on buffer allocation, kernel scheduling, and proper synchronization. (To properly design such a runtime is non-trivial and should happen separately.)  

Which that said, what you said the above sounds great for the long-term goal! But given that the development on the conversion pass just gets started (what you said as 1. Pipeline of the passes in the above), I'd suggest we start smaller and get pieces implemented one by one. :)

### Runtime

Agreed on 2. Runtime machinery. We should have a small library that invokes Vulkan APIs to set up the device, set up the compute pipeline, feed in the data and get the result. Don't worry about extra extensions and capability probing for now; we can just focus on Vulkan 1.0 compute core functionalities. The small library can be exposed with a public API like `runOnVulkan(spirv::ModuleOp, llvm::DenseMap<Descriptor, Content> &data)` or similar so that it can be embedded in testing environment as a oneliner. This interface is brutal of course right now and we'd like to extend it eventually to supply more controls, but that can happen gradually. But I do prefer to auto-deduce information from the SPIR-V module. We should be able to get the interface from `spv.module` by inspecting its global `spv.Variable`s and set up descriptors on Vulkan side so no need to manually set that for each case. (And we can do more gradually in the future.) Serialization library is already there so after inspecting the module, we can serialize it to build the compute pipeline.

But for this, I don't think we can use `mlir::JitRunnerMain`, which goes through LLVM's JITing infra and runs on CPU. There is no path going from SPIR-V to LLVM as of today (although in the future we may eventually have it) and the purpose is to send the SPIR-V binary to Vulkan-capable devices for testing directly. (We are also considering using a software-rendered Vulkan like [SwiftShader](https://github.com/google/swiftshader/blob/master/include/vulkan/vulkan.h) given that it provides better reproducibility when something goes wrong but can happen next too.)

### Conversion

So for 1. Pipeline of the passes, as said before, because it's still under development, we defer it for now. Transforming from some high-level dialect to SPIR-V dialect is separable anyway; from the Vulkan runner's perspective, it only cares about SPIR-V and it does not matter how that SPIR-V is generated. So I'm fine to take in a `spv::module` directly to start. The conversion should be checked by their dedicated FileCheck tests separately anyway.

### Testing harness

Another thing that needs a bit thinking is how we feed in the test data and how we check the result. So from the Vulkan library's perspective, it just takes in a SPIR-V module to construct a compute pipeline, feeds in the data to run it and then returns the result (as shown in the proposed public API in the above: `runOnVulkan(spirv::ModuleOp, llvm::DenseMap<Descriptor, Content> &data)`). We should have a testing harness layered on top of this Vulkan library. Ideally it should be FileCheck so we can have something like:

```
// RUN: mlir-vk-runner --prefix=INPUT-CASE1 ... | FileCheck --check-prefix=OUTPUT-CASE1 ...
// RUN: mlir-vk-runner --prefix=INPUT-CASE2 ... | FileCheck --check-prefix=OUTPUT-CASE2 ...

// INPUT-CASE1: descriptor(0, 1) = [1.0, 2.0, 3.0, 4.0]
// INPUT-CASE1: descriptor(0, 2) = [1.0, 2.0, 3.0, 4.0]
// OUTPUT-CASE1: [2.0, 4.0, 6.0, 8.0]
spv.module {
  ...
}
```

(The above is inspired by [Amber](https://github.com/google/amber/blob/master/tests/cases/compute_mat2x2.vkscript))

But apparently it requires a lot of infra to get to this point! (Parsing floats for setting the contents of descriptors is non-trivial on itself already.)

Another option is actually to use GoogleTest as the testing harness. So we have access to all C++ mechanaries for preparing data and checking the result, which should be much easier. But it does mean that we are having two places for a test: one `.mlir` file containing the module, and another `.cpp` file containing the input data and output data, which should be fine for now to start with.

Thanks again for take on this task! I really appreciate your contributions! Let me know if the above is clear; I can certainly elaborate. Please also feel free to propose alternatives. :)

**Comment by denis0x0D** [comment]
@MaheshRavishankar @antiagainst thanks a lot for the answers.
@antiagainst 

> The small library can be exposed with a public API like runOnVulkan(spirv::ModuleOp, llvm::DenseMap<Descriptor, Content> &data) or similar so that it can be embedded in testing environment as a oneliner. This interface is brutal of course right now and we'd like to extend it eventually to supply more controls, but that can happen gradually. But I do prefer to auto-deduce information from the SPIR-V module. We should be able to get the interface from spv.module by inspecting its global spv.Variables and set up descriptors on Vulkan side so no need to manually set that for each case. (And we can do more gradually in the future.) 


> So I'm fine to take in a spv::module directly to start. The conversion should be checked by their dedicated FileCheck tests separately anyway.


sounds great to me, thanks! I was thinking how split the huge cuda-pipeline and make similiar for Vulkan on the first iteration.

> Another option is actually to use GoogleTest as the testing harness. So we have access to all C++ mechanaries for preparing data and checking the result, which should be much easier. But it does mean that we are having two places for a test: one .mlir file containing the module, and another .cpp file containing the input data and output data, which should be fine for now to start with.

Thanks got it!

> But for this, I don't think we can use mlir::JitRunnerMain, which goes through LLVM's JITing infra and runs on CPU. There is no path going from SPIR-V to LLVM as of today (although in the future we may eventually have it) and the purpose is to send the SPIR-V binary to Vulkan-capable devices for testing directly. 

can you please clarify, I actually, meant the following pipeline, please fix my wrong assumptions:
1. spirv.module with kernel and main function
```
module {
  spv.module "Logical" "VulkanKHR" {
    %0 = spv.Variable bind(0, 0) : !spv.ptr<f32, StorageBuffer>
    %1 = spv.Variable bind(0, 1) : !spv.ptr<!spv.array<12 x f32>, StorageBuffer>
    func @kernel_1(%arg0: !spv.ptr<f32, StorageBuffer>, %arg1: !spv.ptr<!spv.array<12 x f32>, StorageBuffer>) {
      spv.Return
    }
    spv.EntryPoint "GLCompute" @kernel_1, %0, %1 : !spv.ptr<f32, StorageBuffer>, !spv.ptr<!spv.array<12 x f32>, StorageBuffer>
  }
  func @kernel_1(%arg0: f32, %arg1: memref<12xf32, 1>)
  attributes  {gpu.kernel} {
    return
  }
  func @main() {
    %0 = "op"() : () -> f32
    %1 = "op"() : () -> memref<12xf32, 1>
    %c1 = constant 1 : index
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %0, %1) {kernel = @kernel_1} : (index, index, index, index, index, index, f32, memref<12xf32, 1>) -> ()
    return
  }
}
```
2. Splits this spv.module into spirv assembly file or binary and into main function:
2.1.

```
; SPIR-V
; Version: 1.0
; Generator: Khronos; 22
; Bound: 14
; Schema: 0
               OpMemoryModel Logical VulkanKHR
               OpEntryPoint GLCompute %kernel_1 "kernel_1" %3 %8
               OpName %kernel_1 "kernel_1"
               OpDecorate %3 Binding 0
               OpDecorate %3 DescriptorSet 0
               OpDecorate %8 Binding 1
               OpDecorate %8 DescriptorSet 0
      %float = OpTypeFloat 32
%_ptr_StorageBuffer_float = OpTypePointer StorageBuffer %float
        %int = OpTypeInt 32 1
     %int_12 = OpConstant %int 12
%_arr_float_int_12 = OpTypeArray %float %int_12
%_ptr_StorageBuffer__arr_float_int_12 = OpTypePointer StorageBuffer %_arr_float_int_12
       %void = OpTypeVoid
          %9 = OpTypeFunction %void %_ptr_StorageBuffer_float %_ptr_StorageBuffer__arr_float_int_12
          %3 = OpVariable %_ptr_StorageBuffer_float StorageBuffer
          %8 = OpVariable %_ptr_StorageBuffer__arr_float_int_12 StorageBuffer
   %kernel_1 = OpFunction %void None %9
         %12 = OpFunctionParameter %_ptr_StorageBuffer_float
         %13 = OpFunctionParameter %_ptr_StorageBuffer__arr_float_int_12
               OpReturn
               OpFunctionEnd
```
2.2.
```
 func @main() {
    %0 = "op"() : () -> f32
    %1 = "op"() : () -> memref<12xf32, 1>
    %c1 = constant 1 : index
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %0, %1) {kernel = @kernel_1} : (index, index, index, index, index, index, f32, memref<12xf32, 1>) -> ()
    return
  }
```
3. Instruments "main" function with just external calls to like transfer host memory to gpu, launch the kernel and so on and convert to llvm.
```
func @main() {
...
%48 = llvm.call @mcuLaunchKernel(%35, %c8, %c8, %c8, %c8, %c8, %c8, %2, %34, %38, %47    ) : (!llvm<"i8*">, index, index, index, index, index, index, !llvm.i32, !llvm<"i8*">, !llvm<"i8**    ">, !llvm<"i8**">) -> !llvm.i32
...

```
 }

So, this kind a function consumes the pointer to the buffer with spirv binary or filename with spirv binary (reads from file),
and launch it:
```
...
VkShaderModuleCreateInfo shaderModuleCreateInfo = {
 VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO, 0, 0, size, shader_ptr};
 VkShaderModule shader_module;
 vkCreateShaderModule(device, &shaderModuleCreateInfo, 0,
                                            &shader_module);
...
...
vkCmdDispatch(commandBuffer, 1, 1, 1);

...
```

4. Invokes ExecuteEngine. so we have only llvm ir. It has entry point as main function, all external symbols must be resolved by runtime linker.

Is it theoretically possible? Please correct me, I might miss a lot.

Thanks a lot for the comments!

**Comment by antiagainst** [comment]
Ah, I misunderstood your idea previously; I thought you'd like to JITing SPIR-V too. Sorry about that! If you'd actually like just to reuse the `mlir::JitRunnerMain` to run the `main` function (for the Vulkan host code), which [references external functions](https://github.com/tensorflow/mlir/blob/10d0c61266696c4af211ee09669c3242ed3beea4/test/mlir-cuda-runner/gpu-to-cubin.mlir#L29-L30) that got resolved to Vulkan API calls in a [runtime wrapper](https://github.com/tensorflow/mlir/blob/10d0c61266696c4af211ee09669c3242ed3beea4/tools/mlir-cuda-runner/cuda-runtime-wrappers.cpp#L89-L108), then that sounds good to me!

I was initially thinking this can be the next step given that we are still working on the conversion part and it requires it. OTOH, a stand-alone Vulkan library would always be good to have. Because we can make it to just take in a SPIR-V module, it's easier to have it as a first step. But if you'd like to go directly to use `mlir::JitRunnerMain`, that's entirely fine for me! And you are always very welcome to fix anything or add any functionality that might be missing in the current flow! :)

One additional thing, feel free to introduce external functions that suit Vulkan's need. We don't need to strictly follow CUDA side here given that Vulkan APIs are different from CUDA's. Related, regarding the GPU dialect, right now we only have launch ops, but in the future I'd expect we'll experiment with better first-class modelling of other host side concepts like memory management, synchronization, etc. So what's external functions right now might be GPU dialect ops in the future. :)

**Comment by denis0x0D** [comment]
@antiagainst thanks a lot for the answer!

> , a stand-alone Vulkan library would always be good to have. Because we can make it to just take in a SPIR-V module, it's easier to have it as a first step.

Absolytely agree with an implementation step by step. 

> small library can be exposed with a public API like runOnVulkan(spirv::ModuleOp, llvm::DenseMap<Descriptor, Content> &data) or similar so that it can be embedded in testing environment as a oneliner. 

I was kind a stuck to find a starting point for mlir-vulkan-runner this starting point looks great to me!

>  If you'd actually like just to reuse the mlir::JitRunnerMain to run the main function (for the Vulkan host code), which references external functions that got resolved to Vulkan API calls in a runtime wrapper, then that sounds good to me!

Thanks a lot for clarification, it would be nice to reuse this infra in future runtime improvement.

> One additional thing, feel free to introduce external functions that suit Vulkan's need. We don't need to strictly follow CUDA side here given that Vulkan APIs are different from CUDA's. Related, regarding the GPU dialect, right now we only have launch ops, but in the future I'd expect we'll experiment with better first-class modelling of other host side concepts like memory management, synchronization, etc. So what's external functions right now might be GPU dialect ops in the future. :)

Thanks, got it!

Thanks!

**Comment by denis0x0D** [comment]
@antiagainst @MaheshRavishankar I want to share some intermidiate results on this tasks, can you please comment on the followring.
The simple Vulkan runtime is almost done, it does all needed runtime machinery like:
creates vulkan Instance, Device, MemoryBuffers, ShaderModule, DescriptorSetLayoutBindings, DescriptorSetLayoutInfo, PipelineLayout, Pipeline, DescriptorPool, DescriptorSets, CommandsBuffers and submits device Queue. For this moment it works only **with one DescriptorSets.** 
Here is a [commit](https://github.com/denis0x0D/mlir/commit/e779bb932d97581ea45ddd840e38f61deee5ea93)

Also, for this moment it provides two public API functions:
```
extern mlir::LogicalResult
runOnShader(llvm::SmallVectorImpl<char> &binaryShader,
            llvm::DenseMap<Descriptor, VulkanBufferContent> &data,
            const VulkanExecutionContext &);

extern mlir::LogicalResult
runOnModule(mlir::ModuleOp, llvm::DenseMap<Descriptor, VulkanBufferContent> &)
```

The first one function and mlir-vulkan-runner binary are only for current tests purpose. With first one I'm able to run and execute a simple compute shader which multiplies two float type arrays element by element.

The second one consumes spirv.Module, deduces some needed information like LocalSize of ExectuionMode, EntryPoint, verifies variables againts sumbited content like llvm::DenseMap<Descriptor, VulkanBufferContent> (This is not done, but I think it's not a hard one)
Serializes module and submits the serilalized module to VulkanRuntime pipeline same as first API function does.

Here are some problems I've faced with current serialization machinery, please fix me, if I'm wrong somewhere:
1. Currently StructType is not handled by serializer, so I'm not able to serialize a module like this:

```
func @spirvmodule() -> () {
  spv.module "Logical" "VulkanKHR" {
    %0 = spv.Variable bind(0, 0) : !spv.ptr<!spv.struct<!spv.array<4 x f32>>, StorageBuffer>
    func @kernel_1(%arg0: !spv.ptr<!spv.struct<!spv.array<4 x f32>>, StorageBuffer>) {
      %c0 = spv.constant 0: i32
      %c1 = spv.constant 1: i32
      %1 = spv.AccessChain %arg0[%c0, %c1] : !spv.ptr<!spv.struct<!spv.array<4 x f32>>, StorageBuffer>
      %2 = spv.Load "StorageBuffer" %1 : f32
      %3 = spv.FAdd %2, %2 : f32
      spv.Store "StorageBuffer" %1, %3 : f32
      spv.Return
    }
    spv.EntryPoint "GLCompute" @kernel_1, %0 : !spv.ptr<!spv.struct<!spv.array<4 x f32>>, StorageBuffer>
    spv.ExecutionMode @kernel_1 "LocalSizeHint", 1, 1, 1
  }
  return
}
```
This is probably for a reason, because  when serializing 
`   %0 = spv.Variable bind(0, 0) : !spv.ptr<!spv.struct<!spv.array<4 x f32>>, StorageBuffer> `
We have to decorate the struct type with
`OpDecorate %structType BufferBlock`
2. Need to provide `OpCapability` to generate valid compute shader.
3. Serialize arrayType with ArrayStride info like
`OpDecorate %arrayType ArrayStride stride`
4. `spirv-val` tool requires that function must begin with `OpLabel`, so we need this op to test serialized module.

BTW, do someone work on StructType serialization, or I can take a look on it?
Thanks.

**Comment by antiagainst** [comment]
> The simple Vulkan runtime is almost done

This is awesome!

> for this moment it provides two public API functions

Both are reasonable for me. And we can revise as we go. 

> 1. Currently StructType is not handled by serializer

You are right. If you are interested in picking up the task, you are more than welcome! :)

> 2. Need to provide OpCapability to generate valid compute shader.

Yes. I think for now we can just emit `Shader` capability in the serializer specially given it's the only one needed for the simple cases right now. I'm thinking about how to handle capabilities in general; it should come later.

> 3. Serialize arrayType with ArrayStride info like

Actually the `spv.array` type need to be extended to allow layout decorations (like `spv.struct`). After that additionally emitting the `OpDecorate` instruction in the serializer should be easy.

> 4. spirv-val tool requires that function must begin with OpLabel, so we need this op to test serialized module.

Good catch! It's because we are not handling control flows and multiple basic blocks in functions yet. So forgot to emit the `OpLabel` for the first block. That's a bug that should be fixed.

**Comment by MaheshRavishankar** [comment]
Thanks @denis0x0D ! This is really great. I think the best way forward would be to just create a bugs for each thing missing, and we can assign ourselves to it so that we can co-ordinate. Since you have an idea of what you need for the runner, do you mind filing the bugs. Thanks!

**Comment by denis0x0D** [comment]
@antiagainst @MaheshRavishankar thanks for the comments. So, I'm about to create 3 issues,
1. Support serialization for spv.struct.
2. Extend spv.array with layoutinfo and update serialization.
3. Emit OpCapability for spv.Module serialization.

Let me know if it's ok, and which one I can start on.

@antiagainst 

>  It's because we are not handling control flows and multiple basic blocks in functions yet

Please fix me if I'm wrong, but it seems like a blocker for Vulkan runtime. In this case vkCreateShaderModule will return 

> VK_ERROR_INVALID_SHADER_NV

because it applies validation pass on the shader, at least seems like my Vulkan driver does it.

Thanks.

**Comment by antiagainst** [comment]
> I'm about to create 3 issues

SGTM! Please feel free to create them! :) 

> which one I can start on.

Feel free to indicate which one you are interested and I'll assign that one to you.

> Please fix me if I'm wrong, but it seems like a blocker for Vulkan runtime. In this case vkCreateShaderModule will return

Ah, I'm just explaining why it is missing at the current moment. It's definitely something we need to fix. It's just not yet implemented.

**Comment by denis0x0D** [comment]
@antiagainst thanks, got it!

**Comment by MaheshRavishankar** [comment]
FYI, This seems like a blocker for this work as well https://github.com/tensorflow/mlir/issues/68

**Comment by denis0x0D** [comment]
@antiagainst I'm now working on unittests, can you please specify what is better: embedd spv.module into string in text format or create spv.module by builder?
Thanks.

**Comment by denis0x0D** [comment]
@antiagainst thanks for answer, basically I was thinking about simple API and library on the first iteration
```
extern mlir::LogicalResult
runOnModule(mlir::ModuleOp, llvm::DenseMap<Descriptor, VulkanBufferContent> &)
```
and to write a unittest which populates test data, calls that function and evaluates the result.
In this case I can embedd the spv.module into string like
std::string moduleText = "spv.module "Logical" "GLSL450" { ... }" and then parse it, or create it using an API builder, same as for other unittest. I'm not sure what is better?

The mlir-vulkan-runner is just for test puropse, I think I still far away from the level of gpu-to-cubin, sorry about it.
Thanks.

**Comment by denis0x0D** [comment]
@dneto0 thanks a lot to comment on this! Lei already mentioned Amber in this thread, it looks really nice and would be great to use it.
By the way,as far as I understood the final goal of this is to be run it by llvm JIT after convertion pipeline, please fix me if I'm wrong.

Pipeline of the passes:
a. The pass which outlines gpu.launch body to kernel function. (GpuKernelOutLiningPass)
b. The pass which converts GPU dialect into SPIR-V dialect. (GPUToSPIRVPass)
c. The codegen stage which generates spir-v binary from spir-v kernel. (SPIR-V serialization) is similiar to ConvertGPUKernelToCubinPass.
d. The pass which replaces "gpu.launch_function" with call into runtime part, and instruments code with calls to runtime machinery simliar to GpuLanchFunctToCudaCallsPass does.
Instrumented calls go to shared library which manages Vulkan runtime.

And Run it by llvm jit compiler.

@antiagainst can you please comment on this?

**Comment by denis0x0D** [comment]
@joker-eph thanks for comment on this. It was a discussion in this thread at the beggining about implementation details.

> Runtime machinery. We should have a small library that invokes Vulkan APIs to set up the device, set up the compute pipeline, feed in the data and get the result. Don't worry about extra extensions and capability probing for now; we can just focus on Vulkan 1.0 compute core functionalities. The small library can be exposed with a public API like` runOnVulkan(spirv::ModuleOp, llvm::DenseMap<Descriptor, Content> &data`) or similar so that it can be embedded in testing environment as a oneliner. This interface is brutal of course right now and we'd like to extend it eventually to supply more controls, but that can happen gradually. 

> Another option is actually to use GoogleTest as the testing harness. So we have access to all C++ mechanaries for preparing data and checking the result, which should be much easier. But it does mean that we are having two places for a test: one .mlir file containing the module, and another .cpp file containing the input data and output data, which should be fine for now to start with.

BTW, even simple Vulkan runtime pipeline for compute shader is large, it would be nice to split this big task into small ones, but I might be wrong, please fix me in this case.
Thanks.

**Comment by joker-eph** [comment]
Splitting big task into small ones is of course important, but again in general we shy away from C++ unit-tests when not strictly needed. I'd even say that we organize our work and design our software so that we can test it as much as possible without these. 
Here I'd like to understand more why the approach taken to bring up the CUDA testing can't be replicated?  Basically the plan you initially introduced here: https://github.com/tensorflow/mlir/issues/60#issuecomment-517805953 looked perfect to me (I read this almost exactly as how the CUDA pipeline was brought up) and so I didn't follow closely after.


> But for this, I don't think we can use mlir::JitRunnerMain, which goes through LLVM's JITing infra and runs on CPU. There is no path going from SPIR-V to LLVM as of today (although in the future we may eventually have it) and the purpose is to send the SPIR-V binary to Vulkan-capable devices for testing directly

This is missing the point of the JITRunner I believe: the JIT runner is more about the host code (the code involving `gpu.launch`) than the SPIR-V module.
The alternative to the JITRunner would be to emit LLVM IR, and then separately call clang on the generated file and execute the resulting binary. The JIT's purpose was to reduce the amount of step and tools involved (since we have a compiler/linker/loader available as a library, why not package it all?!).

I'll try to sync up directly with @antiagainst tomorrow, I wouldn't want to hold up anything too much here, I am super excited to see the mlir-vulkan-runner landing :)

**Comment by antiagainst** [comment]
Sorry for the late reply!

@denis0x0D:

> Pipeline of the passes:

SGTM to have this as initial testing infra! There is one thing that concerns me though: we are not directly testing the SPIR-V dialect per se. This flow bundles the conversion from GPU+std dialects to the SPIR-V dialect. And the test module is not even expressed via the SPIR-V dialect. So it feels more like testing GPU+std works on Vulkan... However, given that we are still evolving the GPU dialect, I think we may sort this out eventually. The test infra developed here will be useful either way so I'm fine with it. :)  

@joker-eph: 

> This is missing the point of the JITRunner I believe: the JIT runner is more about the host code (the code involving gpu.launch) than the SPIR-V module.

Right. I misunderstood the idea previously and explained further at https://github.com/tensorflow/mlir/issues/60#issuecomment-517927732. Quoting myself and comment it more:

> Ah, I misunderstood your idea previously; I thought you'd like to JITing SPIR-V too. Sorry about that! If you'd actually like just to reuse the mlir::JitRunnerMain to run the main function (for the Vulkan host code), which references external functions that got resolved to Vulkan API calls in a runtime wrapper, then that sounds good to me!

> I was initially thinking this can be the next step given that we are still working on the conversion part and it requires it. OTOH, a stand-alone Vulkan library would always be good to have. Because we can make it to just take in a SPIR-V module, it's easier to have it as a first step. But if you'd like to go directly to use mlir::JitRunnerMain, that's entirely fine for me! And you are always very welcome to fix anything or add any functionality that might be missing in the current flow! :)

Please note this was commented a month ago. Back then we don't have the conversion to SPIR-V dialect from the GPU dialect. So I was suggesting an alternative approach via GoogleTest, because it was immediately doable and not blocked by the conversion progress.

**Comment by denis0x0D** [comment]
@antiagainst @joker-eph thanks a lot for the asnwers.
@antiagainst:

> There is one thing that concerns me though: we are not directly testing the SPIR-V dialect per se. This flow bundles the conversion from GPU+std dialects to the SPIR-V dialect. And the test module is not even expressed via the SPIR-V dialect. So it feels more like testing GPU+std works on Vulkan... However, given that we are still evolving the GPU dialect, I think we may sort this out eventually.

I could missunderstand something, please fix me, but is it possible, for current moment, to introduce initial version as shared lib and one API call runOnVulkan which consumes the spirv::ModuleOp and one small unit test, which will be deleted in the future, I hope.
My concern here is that even simple computation pipeline for vulkan is large, at least it requires to create:

1.VkInstance; 2.VkDevice; 3.VkShaderModule; 4.VkDescriptorSetLayoutBinding; 5.VkPipelineLayout;   
6. VkPipeline; 7.VkDescriptorPool; 8.VkDescriptorSet; 9.VkCommandPool; 10.VkCommandBuffer;         11.VkQueue and so on.

and it needs to be carefully implemented and reviewed at first.

IMHO, it seems like to be able to run somehting like this

```
module {
  spv.module "Logical" "VulkanKHR" {
    %0 = spv.Variable bind(0, 0) : !spv.ptr<f32, StorageBuffer>
    %1 = spv.Variable bind(0, 1) : !spv.ptr<!spv.array<12 x f32>, StorageBuffer>
    func @kernel_1(%arg0: !spv.ptr<f32, StorageBuffer>, %arg1: !spv.ptr<!spv.array<12 x f32>, StorageBuffer>) {
      spv.Return
    }
    spv.EntryPoint "GLCompute" @kernel_1, %0, %1 : !spv.ptr<f32, StorageBuffer>, !spv.ptr<!spv.array<12 x f32>, StorageBuffer>
  }
  func @kernel_1(%arg0: f32, %arg1: memref<12xf32, 1>)
  attributes  {gpu.kernel} {
    return
  }
  func @main() {
    %0 = "op"() : () -> f32
    %1 = "op"() : () -> memref<12xf32, 1>
    %c1 = constant 1 : index
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %0, %1) {kernel = @kernel_1} : (index, index, index, index, index, index, f32, memref<12xf32, 1>) -> ()
    return
  }
}
```
in JIT manner requires to rewrite current JITRunner. 
We need to split actual kernel and "main" function, serialize kernel into spir-v binary and then somehow pass a pointer to the launch fuction alongside with creation of all needed vulkan runtime machinery.  I belive it should be properly designed at first in RFC style. 
Please fix me if I'm missing something here.
Thanks.

**Comment by antiagainst** [comment]
>  for current moment, to introduce initial version as shared lib and one API call runOnVulkan which consumes the spirv::ModuleOp and one small unit test, which will be deleted in the future, I hope. My concern here is that even simple computation pipeline for vulkan is large

SGTM to have `runOnVulkan` taking a `spv.module` and test it with GoogleTest and then later replace it with the runner. I'm also okay if you'd like to directly implement the full runner (now we have the conversion there so it's possible). Given that you are implementing this, I'd say whatever way makes your implementation easier. :) 

> in JIT manner requires to rewrite current JITRunner.

I think the JitRunner is configurable? It takes a list of transformations to apply. We just don't need all the transformations going from GPU+std to SPIR-V and we just start from serializing SPIR-V directly? I mean, something like

```
module {
  spv.module ...
  func @main() {
    call @vkPopulateBuffer() ...
    // ...
    call @vkLaunchShader() ...
    // ...
 }

// extern @vkPopulateBuffer
// extern @vkLaunchShader
// ...
```

But I guess you can ignore my concern for now. After discussing with Mehdi offline, he pointed out that if we do this, it is essentially just testing SPIR-V serialization from MLIR's perspective. The runner is expected to do integration-style tasks. So it's actually more suitable to start with some higher level dialects and checks that higher level representation is able to be run. (The runner itself is configurable so it can take in whatever dialect as long as we plugin the correct set of transformations. So it does not preclude us from using it to run SPIR-V dialect on Vulkan devices for tasks like profiling. But using it for testing the SPIR-V dialect might not be that useful.) I think Mehdi will comment later regarding general thoughts on runners in MLIR. :)

**Comment by denis0x0D** [comment]
@antiagainst thanks a lot for the detailed answer!

**Comment by joker-eph** [comment]
> IMHO, it seems like to be able to run somehting like this
> [...]
> in JIT manner requires to rewrite current JITRunner. We need to split actual kernel and "main" function, serialize kernel into spir-v binary and then somehow pass a pointer to the launch fuction alongside with creation of all needed vulkan runtime machinery. I belive it should be properly designed at first in RFC style. Please fix me if I'm missing something here.

Right, but that's what I mentioned before when I wrote:  "I'd even say that we organize our work and design our software so that we can test it as much as possible without these".
You can look at how the mlir-cuda-runner was landed for example.

I am not really sure how to consider any runtime API independently of the lowering. It seems that the API that `gpu.launch_func` needs to be lowered to is probably the first thing to define.
Then I would expect to see the equivalent to these: https://cs.corp.google.com/piper///depot/google3/third_party/llvm/llvm/projects/google_mlir/test/Conversion/GPUToCUDA/?g=0
Having these conversions are prerequisite to me.


> I think Mehdi will comment later regarding general thoughts on runners in MLIR. :)

After discussing with Lei, something wasn't maybe clear to everyone: to test shaders written in SPIR-V on a Vulkan implementation, it seems like projects like Amber are already fitting the role. Also, this is not very interesting from an MLIR point of view since this would only exercise the SPIR-V serializer.
On the other hand, the reason to have runners in MLIR is to provide a path for integration test focused on testing the layers *above* the SPIR-V dialect. For example, the CPU runner is used for some Linalg integration tests: https://github.com/tensorflow/mlir/blob/master/test/mlir-cpu-runner/linalg_integration_test.mlir
Similarly, in TensorFlow we would be interested to use the runner to be able to validate the pipeline, being able to write:

```
// RUN: tf-gpu-runner %s -init-value "[1, 2, 3], [2, 3, 4]"
func @main(%a : tensor<3xf32>, %b : tensor<3xf32>) {
  %mul = "tf.Dot"(%a, %b) { "device" : "GPU0"} : (tensor<3xf32>, tensor<3xf32>) -> tensor<f32>
  "tf.Print"(%mul)
// CHECK: "[20]"
}
```

All the interesting parts of the integration test happen *before* you reach the final SPIR-V assembly.
Also this is not something you would want to FileCheck on the SPIR-V assembly as there are many parts of the lowering that could impact the actual CodeGen. The contract of the integration test is not on the actual SPIR-V IR emitted.

The runner purpose is to support this kind of integration testing. It also provides an easy bring up for getting something end-to-end running, for instance adding a GPU path to the Toy tutorial should be "easy" as soon as you have the runner infrastructure available.

**Comment by denis0x0D** [comment]
@joker-eph thansk for the asnwer! If I got you right, the current blocker for Vulkan runtime is:
- The pass which replaces "gpu.launch_function" with call into runtime part, and instruments code with calls to runtime machinery simliar to GpuLanchFunctToCudaCallsPass does. The result of the pass must be something like Lei mentioned:
```
module {
  spv.module ...
  func @main() {
    call @vkPopulateBuffer() ...
    // ...
    call @vkLaunchShader() ...
    // ...
 }

// extern @vkPopulateBuffer
// extern @vkLaunchShader
// ...
```
The extern functions are just wrappers around the actual Vulkan runtime.

BTW, as far as I understood the concept behind MLIR, in addition to unification, it closes the gap between high level structure such an AST and lowe level IR suitable for codegeneration.
Most of the current dialects reuse llvm infra to generate the target code directly or by convertion to HLO dialect (XLA) which again reuses llvm for codegen. In this case SPIR-V dialect looks an exeptional from other; it has serialization machinery which produces binary code, yes it's not a machine code, by still can be directly passing to the driver. And as far as I understood, please fix me if I'm wrong, the MLIR must provide some Vulkan runtime not even for test purpose. For example if I have some representaion in TF dialect -> GPU dialect -> SPIR-V dialec -> SPIRV assembly, what the actual purpose of SPIR-V assembly from TF dialect without runtime? Please fix me if I'm wrong, I can miss a lot.
Thanks.

**Comment by joker-eph** [comment]
> In this case SPIR-V dialect looks an exeptional from other; it has serialization machinery which produces binary code, yes it's not a machine code, by still can be directly passing to the driver.

This does not seem different from the PTX generated by LLVM that can be present in the cubin and passed to the Nvidia driver? The fact that LLVM is involved to emit the binary blob and embed it in the IR or another serialization mechanism for SPIR-V seems fairly opaque to me here.

Your edit seems right to me that it is likely that a final users will likely want to integrate in their own runtime environment, either by customizing the lowering of the GPU dialect to their own APIs, or by rewriting/forwarding the calls to vkPopulateBuffer/vkLaunchShader in your example above.
(or other strategy TBD...)

**Comment by denis0x0D** [comment]
@joker-eph thanks for the answer, it's clear now for me.
@antiagainst @MaheshRavishankar can you please share, do you have any plans to implement the pass which is similar to this https://github.com/tensorflow/mlir/blob/master/lib/Conversion/GPUToCUDA/ConvertLaunchFuncToCudaCalls.cpp#L289
Thanks.

**Comment by antiagainst** [comment]
>  the actual runtime implentation is the responsobility of the MLIR users

Yes some users would like to have their own runtime. They can use the SPIR-V lowering path in a isolated way: the user partitions the computation graph by itself, extracts suitable subgraphs and feeds it to the SPIR-V lowering path, and get the final SPIR-V blob and uses it to create a shader module. The MLIR system has no interaction with the runtime side whatsoever; it only touches kernel compilation. But with the GPU dialect, we are thinking maybe we can do more than that by introducing ops for runtime API calls (`gpu.launch` is already one). We can add other ops for memory allocation and synchronization, so the runtime side can also be represented in MLIR and we can use compiler techniques to optimize here (like analyzing buffer usages, removing unused buffers, combining buffers, etc.) and generate C++ API calls from them. These are all interesting ideas worth exploring. This is what I meant by saying the GPU dialect is still evolving and it's a bit unclear right now how the future would look like. :) But I won't base mlir-vulkan-runner's development on the above. Your approach of taking small steps to support what we already have is an excellent one. We will need these building blocks anyway so it's awesome to have them early and we can certainly adjust accordingly. ;)

> Do you have any plans to implement the pass which is similar to this 

We don't have immediate plans for that right now. So if you are interested to make it happen, please feel free and thanks a lot! :)

**Comment by denis0x0D** [comment]
@antiagainst thanks a lot for the answer!

> We can add other ops for memory allocation and synchronization, so the runtime side can also be represented in MLIR and we can use compiler techniques to optimize here (like analyzing buffer usages, removing unused buffers, combining buffers, etc.) and generate C++ API calls from them.

This is interesting to know!
BTW, Is it ok, if I'll create a new issue for the task related to that pass?

**Comment by antiagainst** [comment]
Sure!

**Comment by jon-chuang** [comment]
Is there any relationship between the MLIR spir-v/cuda dialects and the current LLVM CUDA backend?

**Comment by antiagainst** [comment]
Hi @jon-chuang, please note that MLIR now is already in LLVM monorepo so issues on this repo is not maintained anymore.

Coming to your question, there is no relationship between mlir-vukan-runner with LLVM CUDA backend. Vulkan is a separate API from CUDA and SPIR-V in MLIR does not rely on LLVM IR itself.

Closing this given mlir-vulkan-runner is already landed.
